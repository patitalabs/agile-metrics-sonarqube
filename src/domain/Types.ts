import { SaCollectorConfig, SaMetricItem } from './static-analysis/Types';

export interface TeamMetricsRequest {
  shouldUpdateEntries: boolean;
  config: any;
}

export interface CoreMetricsService {
  publish(entries: any, shouldReplaceEntries: boolean);
}

export interface SaCollectorService {
  fetch(collectorConfig: SaCollectorConfig): Promise<SaMetricItem[]>;
}
