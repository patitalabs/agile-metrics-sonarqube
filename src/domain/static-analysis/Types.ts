export interface SaMetricItem {
  id: string;
  dataType: string;
  createdAt: Date;
  teamName: string;
  projectName: string;
  alertStatus: string;
  qualityGateDetails: string;
  bugs: number;
  newBugs: number;
  reliabilityRating: number;
  newReliabilityRating: number;
  vulnerabilities: number;
  newVulnerabilities: number;
  securityRating: number;
  newSecurityRating: number;
  codeSmells: number;
  newCodeSmells: number;
  sqaleRating: number;
  newMaintainabilityRating: number;
  sqaleIndex: number;
  newTechnicalDebt: string;
  coverage: number;
  newCoverage: number;
  newLinesToCover: number;
  tests: number;
  duplicatedLinesDensity: number;
  newDuplicatedLinesDensity: number;
  duplicatedBlocks: number;
  ncloc: number;
  nclocLanguageDistribution: string;
  newLines: number;
  version: string;
}

export class SaCollectorConfig {
  projectName: string;
  teamName: string;
  since: string;
  until?: string;

  constructor({ projectName, since, until = null, teamName }) {
    this.projectName = projectName;
    this.since = since;
    this.until = until;
    this.teamName = teamName;
  }
}
