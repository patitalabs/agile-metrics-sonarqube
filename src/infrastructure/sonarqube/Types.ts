export class ProjectMetrics {
  createdAt: Date;
  teamName: string;
  projectName: string;
  alertStatus: string;
  qualityGateDetails: string;
  bugs: number;
  newBugs: number;
  reliabilityRating: number;
  newReliabilityRating: number;
  vulnerabilities: number;
  newVulnerabilities: number;
  securityRating: number;
  newSecurityRating: number;
  codeSmells: number;
  newCodeSmells: number;
  sqaleRating: number;
  newMaintainabilityRating: number;
  sqaleIndex: number;
  newTechnicalDebt: string;
  coverage: number;
  newCoverage: number;
  newLinesToCover: number;
  tests: number;
  duplicatedLinesDensity: number;
  newDuplicatedLinesDensity: number;
  duplicatedBlocks: number;
  ncloc: number;
  nclocLanguageDistribution: string;
  newLines: number;
  version: string;
}

export interface SonarConfig {
  projectName: string;
  teamName: string;
  since: string;
  until?: string;
}

export interface SonarService {
  projectMetrics(sonarConfig: SonarConfig): Promise<ProjectMetrics[]>;
}

export interface SonarClient {
  getData(url: string): Promise<any>;
}

export interface SonarRepository {
  projectMetrics(sonarConfig: SonarConfig): Promise<ProjectMetrics[]>;
}
