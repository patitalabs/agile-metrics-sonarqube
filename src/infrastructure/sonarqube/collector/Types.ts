export interface SonarMetricItem {
  projectName: string;
  alertStatus: string;
  qualityGateDetails: string;
  bugs: number;
  newBugs: number;
  reliabilityRating: number;
  newReliabilityRating: number;
  vulnerabilities: number;
  newVulnerabilities: number;
  securityRating: number;
  newSecurityRating: number;
  codeSmells: number;
  newCodeSmells: number;
  sqaleRating: number;
  newMaintainabilityRating: number;
  sqaleIndex: number;
  newTechnicalDebt: string;
  coverage: number;
  newCoverage: number;
  newLinesToCover: number;
  tests: number;
  duplicatedLinesDensity: number;
  newDuplicatedLinesDensity: number;
  duplicatedBlocks: number;
  ncloc: number;
  nclocLanguageDistribution: string;
  newLines: number;
  version: string;
}

export class StaticAnalysisCollectorConfig {
  projectName: string;
  teamName: string;
  since: string;
  until?: string;

  constructor({ projectName, since, until = null, teamName }) {
    this.projectName = projectName;
    this.since = since;
    this.until = until;
    this.teamName = teamName;
  }
}
