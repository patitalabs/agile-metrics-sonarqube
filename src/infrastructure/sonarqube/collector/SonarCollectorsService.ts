import { SonarMetricConverter } from './SonarMetricConverter';
import { SonarService } from '../Types';
import { SaCollectorService } from '../../../domain/Types';
import {
  SaCollectorConfig,
  SaMetricItem,
} from '../../../domain/static-analysis/Types';

export class SonarCollectorsService implements SaCollectorService {
  constructor(private readonly sonarService: SonarService) {}

  public async fetch(
    saCollectorConfig: SaCollectorConfig
  ): Promise<SaMetricItem[]> {
    const projectMetrics = await this.sonarService.projectMetrics({
      teamName: saCollectorConfig.teamName,
      projectName: saCollectorConfig.projectName,
      since: saCollectorConfig.since,
    });

    return projectMetrics.map((metricItem) =>
      SonarMetricConverter.toMetricItem(metricItem)
    );
  }
}
