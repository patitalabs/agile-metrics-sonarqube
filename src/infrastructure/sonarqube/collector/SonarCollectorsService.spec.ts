import { StaticAnalysisCollectorConfig } from './Types';
import { SonarCollectorsService } from './SonarCollectorsService';
import { ProjectMetrics, SonarConfig, SonarService } from '../Types';

function testProjectMetric(): ProjectMetrics {
  return {
    createdAt: new Date('2018-12-20'),
    teamName: 'someTeamName',
    projectName: 'the project',
    alertStatus: 'OK',
    qualityGateDetails: 'Passed',
    bugs: 1,
    newBugs: 2,
    reliabilityRating: 3,
    newReliabilityRating: 2,
    vulnerabilities: 8,
    newVulnerabilities: 1,
    securityRating: 1,
    newSecurityRating: 1,
    codeSmells: 30,
    newCodeSmells: 3,
    sqaleRating: 1,
    newMaintainabilityRating: 2,
    sqaleIndex: 1,
    newTechnicalDebt: 'technical debt text',
    coverage: 60,
    newCoverage: 60,
    newLinesToCover: 50,
    tests: 20,
    duplicatedLinesDensity: 30,
    newDuplicatedLinesDensity: 10,
    duplicatedBlocks: 10,
    ncloc: 40,
    nclocLanguageDistribution: 'lang distribution',
    newLines: 20,
    version: '2.5',
  };
}

describe('SonarCollectorsService', () => {
  const sonarService: SonarService = {
    projectMetrics: (sonarConfig: SonarConfig): Promise<ProjectMetrics[]> =>
      Promise.resolve([testProjectMetric()]),
  };

  const sonarCollectorsService: SonarCollectorsService =
    new SonarCollectorsService(sonarService);

  it('should fetch sonarMetrics', async () => {
    const staticAnalysisCollectorConfig: StaticAnalysisCollectorConfig =
      new StaticAnalysisCollectorConfig({
        projectName: 'someProject',
        since: '2018-11-20',
        until: '2020-11-20',
        teamName: 'someTeamName',
      });

    const data = await sonarCollectorsService.fetch(
      staticAnalysisCollectorConfig
    );
    expect(data).toMatchSnapshot();
  });
});
