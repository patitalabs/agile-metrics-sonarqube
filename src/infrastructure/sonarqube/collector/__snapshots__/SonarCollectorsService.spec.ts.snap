// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SonarCollectorsService should fetch sonarMetrics 1`] = `
Array [
  Object {
    "alertStatus": "OK",
    "bugs": 1,
    "codeSmells": 30,
    "coverage": 60,
    "createdAt": 2018-12-20T00:00:00.000Z,
    "dataType": "SAT",
    "duplicatedBlocks": 10,
    "duplicatedLinesDensity": 30,
    "id": "1caedba1ffbcccc1e6f9ecac86107d41c13e9426b9115582e119554c53c69d37",
    "ncloc": 40,
    "nclocLanguageDistribution": "lang distribution",
    "newBugs": 2,
    "newCodeSmells": 3,
    "newCoverage": 60,
    "newDuplicatedLinesDensity": 10,
    "newLines": 20,
    "newLinesToCover": 50,
    "newMaintainabilityRating": 2,
    "newReliabilityRating": 2,
    "newSecurityRating": 1,
    "newTechnicalDebt": "technical debt text",
    "newVulnerabilities": 1,
    "projectName": "the project",
    "qualityGateDetails": "Passed",
    "reliabilityRating": 3,
    "securityRating": 1,
    "sqaleIndex": 1,
    "sqaleRating": 1,
    "teamName": "someTeamName",
    "tests": 20,
    "version": "2.5",
    "vulnerabilities": 8,
  },
]
`;
